source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -0,0 +1,431 @@
+      Http::Utility::QueryParams query_parameters = Http::Utility::parseQueryString(path_str);
+
+      const auto state =
+          Http::Utility::PercentEncoding::decode(query_parameters.at(queryParamsState()));
+      Http::Utility::Url state_url;
+      if (!state_url.initialize(state, false)) {
+        sendUnauthorizedResponse();
+        return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+      }
+      // Avoid infinite redirect storm
+      if (absl::StartsWith(state_url.pathAndQueryParams(), config_->callbackPath())) {
+        sendUnauthorizedResponse();
+        return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+      }
+      Http::ResponseHeaderMapPtr response_headers{
+          Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
+              {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))},
+               {Http::Headers::get().Location, state}})};
+      decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+    }
+
+    // Continue on with the filter stack.
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // If a bearer token is supplied as a header or param, we ingest it here and kick off the
+  // user resolution immediately. Note this comes after HMAC validation, so technically this
+  // header is sanitized in a way, as the validation check forces the correct Bearer Cookie value.
+  access_token_ = extractAccessToken(headers);
+  if (!access_token_.empty()) {
+    found_bearer_token_ = true;
+    request_headers_ = &headers;
+    finishFlow();
+
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // If no access token and this isn't the callback URI, redirect to acquire credentials.
+  //
+  // The following conditional could be replaced with a regex pattern-match,
+  // if we're concerned about strict matching against the callback path.
+  if (!absl::StartsWith(path_str, config_->callbackPath())) {
+    Http::ResponseHeaderMapPtr response_headers{Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
+        {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))}})};
+
+    // Construct the correct scheme. We default to https since this is a requirement for OAuth to
+    // succeed. However, if a downstream client explicitly declares the "http" scheme for whatever
+    // reason, we also use "http" when constructing our redirect uri to the authorization server.
+    auto scheme = Http::Headers::get().SchemeValues.Https;
+
+    const auto* scheme_header = headers.Scheme();
+    if ((scheme_header != nullptr &&
+         scheme_header->value().getStringView() == Http::Headers::get().SchemeValues.Http)) {
+      scheme = Http::Headers::get().SchemeValues.Http;
+    }
+
+    const std::string base_path = absl::StrCat(scheme, "://", host_);
+    const std::string callback_path = base_path + config_->callbackPath();
+    const std::string state_path = absl::StrCat(base_path, headers.Path()->value().getStringView());
```

Comment:
This is where that `callback_path` is used in multiple ways (matching incoming request and constructing redirect response).

And for the redirect_uri construction here:

The spec strongly encourage usage of TLS: https://tools.ietf.org/html/rfc6749#section-3.1.2.1
Inferring HTTP/HTTPS from scheme isn't a good way, if this is deployed behind a L7 LB or other Envoy, we should look at `x-forwarded-proto` or other things.

It would be better to provide a configurable `redirect_uri` which specified full path, utilize what our header formatter is doing. so it is flexible to specify `%REQ(:SCHEME)://%REQ(:AUTHORITY)/whatever_redirect_path` to match the behavior here, or use a fixed `https://example.com/oauth_callback` to enforce https and fixed host.

Reply:
I like the idea of providing a header formatted string for the added flexibility, though how would that be reconciled with the check to see we're handling the callback path? I think it opens up for constructing callback paths that might be different when evaluating the header format string when constructing the redirect path vs when the callback is being handled. 

We could just document this and warn about being too creative with the callback URL, and then evaluating the header format string twice and checking for full URI equality. 

Alternatively it might be possible to figure out a way to avoid checking the request to see if it matches the callback. For example, we could look for the expected query parameters and if they exist we assume we're on the callback path.

Thoughts?

Reply:
Or just split what we do match vs what we create. That's also address mismatch when a front proxy/LB does prefix rewrite etc.

But yeah even with that, we definitely need to warn about being too creative about callback URI construction.